### PR TITLE
[BuildConfigGen] Update ways nvm being installed

### PR DESCRIPTION
**Pull request by qianz2**
**Task name**: BuildConfigGen

**Description**: 
Update ways nvm being installed. 
Build default tasks and node 20 tasks seperately.
Update tsc 5.1.6 for node 20 tasks.
Revert tsc 4.0.8 to 4.0.2 on root folder.

**Documentation changes required:** (Y/N) <Please mark if documentation changes are required>

**Added unit tests:** (Y/N) <Please mark if unit tests were added or updated according changes>

**Attached related issue:** (Y/N) <Please add link to related issue here>

**Checklist**:
- [ ] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [ ] Checked that applied changes work as expected

### CODE REVIEW

File: make.js [hunk 1]
```diff
@@ -177,222 +177,59 @@ CLI.gendocs = function() {
...
+    if (nodeVersion == "Node20") {
+        ensureTool('node', '--version', 'v20.3.1');
+        allTasksNode20.forEach(buildTasks);
+    } else if (nodeVersion == "Default") {
+        if (allTasksNode20) {
+            console.log(
+            "==========================================\n" + 
+            "IMPORTANT NOTE: There are additional tasks that need to be build with a different node configuration. \n" + 
+            "Unfortunately, we cannot switch node versions while running make.js.  Run the following commands to continue: \n" +
+            "nvm use 20.3.1\n" +
```
Comment (merlynomsft) [inline make.js:226]: >20.3.1 [](http://example.com/codeflow?start=21&length=6)

we are repeating versions #s a few places -- would be good to set a variable

File: make.js [hunk 2]
```diff
@@ -177,222 +177,59 @@ CLI.gendocs = function() {
...
+            "node make.js build --task taskName --node Node20\n" +
```
Comment (merlynomsft) [inline make.js:227]: >make.js build --task taskName --node Node20\ [](http://example.com/codeflow?start=18&length=44)

[not blocking, could do in another pr] it would be good to put in the exact command line to run, including the user's actual args so they can actually copy+paste

**Comment by merlynomsft** [comment]
> # Clean

it would be good to temporarily push and revert a change a task to ensure the pipelines execute successfully

---
Refers to: ci/build-all-steps.yml:6 in d202d91. [](commit_id = d202d9155e65cdf8100707a154e92753438595ff, deletion_comment = False)

**Comment by merlynomsft** [comment]
> CLI.test = function(/** @type {{ suite: string; node: string; task: string }} */ argv) {

I'm thinking we'll need to make changes to test to.  It's fine to leave it for another PR if test isn't blocked for any task, but we should check

---
Refers to: make.js:437 in d202d91. [](commit_id = d202d9155e65cdf8100707a154e92753438595ff, deletion_comment = False)

File: make.js [hunk 3]
```diff
@@ -402,6 +239,195 @@ CLI.serverBuild = function() {
...
+              rm('-Rf', taskNodeModulesPath);
+          }
+
+          const taskTestsNodeModulesPath = path.join(taskPath, 'Tests', 'node_modules');
+
+          if (fs.existsSync(taskTestsNodeModulesPath)) {
+              console.log('\n> removing task tests node modules');
+              rm('-Rf', taskTestsNodeModulesPath);
+          }
+      }
```
Comment (merlynomsft) [inline make.js:428]: it looks like indent is one level too deep here?

File: make.js [hunk 4]
```diff
@@ -402,6 +239,195 @@ CLI.serverBuild = function() {
     banner('Build successful', true);
 }
 
+function buildTasks(taskName) {
```
Comment (merlynomsft) [inline make.js:242]: >buildTasks [](http://example.com/codeflow?start=9&length=10)

I think this should be "buildTask" instead of "buildsTasks"?  I only see the function building one task

File: make-util.js [hunk 5]
```diff
@@ -16,7 +16,7 @@ var downloadPath = path.join(__dirname, '_download');
 // list of .NET culture names
 var cultureNames = ['cs', 'de', 'es', 'fr', 'it', 'ja', 'ko', 'pl', 'pt-BR', 'ru', 'tr', 'zh-Hans', 'zh-Hant'];
 
-var allowedTypescriptVersions = ['4.0.2', '4.0.8'];
+var allowedTypescriptVersions = ['4.0.2', '4.0.8', '5.1.6'];
```
Comment (merlynomsft) [inline make-util.js:19]: >4.0.2', '4.0.8 [](http://example.com/codeflow?start=34&length=14)

[not blocking, another pr] as mentioned let's stick with 4.0.2 if we can

File: make-util.js [hunk 6]
```diff
@@ -329,8 +329,32 @@ var ensureTool = function (name, versionArgs, validate) {
 }
 exports.ensureTool = ensureTool;
 
+var ensureNvmInstalled = function () {
+    console.log('nvm tool: ');
```
Comment (merlynomsft) [inline make-util.js:333]: >console [](http://example.com/codeflow?start=4&length=7)

[not blocking, can handle in another pr]  we should check for the following: (copy + paste from @ozanaydinMSFT 's comment offline):  

Also; in local environment; developers should update their nvm versions to latest (1.1.11); because "nvm use 20" command is giving error with previous versions

File: make.js [hunk 7]
```diff
@@ -402,6 +239,195 @@ CLI.serverBuild = function() {
...
+      let isGeneratedTask = false;
+      banner('Building: ' + taskName);
+      const removeNodeModules = taskList.length > 1;
```
Comment (merlynomsft) [inline make.js:245]: >removeNodeModules [](http://example.com/codeflow?start=12&length=17)

I think we need to check the filtered list for this, not "taskList", maybe pass it in as an arg?

**Comment by qianz2** [comment]
@merlynomsft , I addressed most of your feedbacks except a few ones, I put it here and will update them in another PR:

- NVM version. 1.1.11 latest is recommended. We can put it into self-documenting notes. (I don't have issues when run nvm install 20 using nvm version < 1.1 on windows, not sure if it's a platform specific requirement.)

- We'll ensure CLI.test will test node default and node20 tasks separately as well. So far I can see an argv that allows `--node` version passed into it. I'll test it locally and update if necessary.

A testing [build](https://dev.azure.com/mseng/PipelineTools/_build/results?buildId=21094586&view=logs&j=55b94f1c-9d32-5114-bc90-83768ebbe0df) was run, pipelines execute `install node 20.x` and `build tasks Node20` jobs successfully.

File: make.js [hunk 8]
```diff
@@ -177,229 +180,255 @@ CLI.gendocs = function() {
...
+        ensureTool('node', '--version', `v${node20Version}`);
+        allTasksNode20.forEach(taskName => buildTask(taskName, allTasksNode20.length));
+    } else if (nodeVersion == "Default") {
+        if (allTasksNode20) {
+            console.log(
+            "==========================================\n" + 
+            "IMPORTANT NOTE: There are additional tasks that need to be build with a different node configuration. \n" + 
+            "Unfortunately, we cannot switch node versions while running make.js.  Run the following commands to continue: \n" +
+            `nvm use ${node20Version}\n` +
+            `node make.js build --task ${argv.task} --node Node20\n` +
```
Comment (merlynomsft) [inline make.js:230]: >--task ${argv.task} [](http://example.com/codeflow?start=32&length=19)

[not blocking, may handle in another pr] this doesn't handle when --task isn't specified (all tasks built)

File: package.json [hunk 9]
```diff
@@ -45,7 +45,7 @@
     "shelljs": "^0.8.5",
     "sync-request": "4.1.0",
     "typed-rest-client": "^1.8.9",
-    "typescript": "4.0.8",
+    "typescript": "^4.0.2",
```
Comment (merlynomsft) [inline package.json:48]: >^ [](http://example.com/codeflow?start=19&length=1)

no ^

**Review by merlynomsft — APPROVED** [review]
:shipit:
